### ISSUE TITLE
qvm-clone across-pool performance - non-optimal dd invocation with default block size

### ISSUE THREAD

**Issue by brendanhoar** [QubesOS/qubes-issues#5134]
**Qubes OS version**
R4.0x w/ current-testing

**Affected component(s) or functionality**
qvm-clone

**Brief summary**
qvm-clone invokes dd with conv=sparse, if=XXX, of=YYY and status=none arguments, but without bs= argument (perhaps only during cross-pool clones?). This leads to a default blocksize of 512 bytes, which severely impacts IO.

**To Reproduce**
% qvm-clone -P ephemeral-test-thinpool whonix-ws-14-dvm whonix-ws-14-dvm-ephemeral # clones VM from default pool to custom pool

This took several minutes during the copy process on an SSD, and the only storage being cloned was a single, nearly empty 16GB "private" thin volume (as per qvm-clone output and verified via post-run sudo lvs). 

From the blinkenlights, the amount of actual IO was small.

**Expected behavior**
The dd portion of qvm-clone would use a performance block size argument, somehwere in the 32KB to 1MB range (tuning is needed). In basic tests on an SSD-based system with thin pools of 64KB-sized clusters, 128KB has appeared to be approximately correct in unrelated dd experiments.

**Actual behavior**
dd portion of qvm-clone uses default block size, which is 512 bytes

**Additional context**
Working on proof-of-concept ephemeral disposable VMs using lvm on top of dm-crypt on top of lvm on top of luks. I was expecting the launch time to be high...but not that high!

**Comment by brendanhoar** [comment]
Looks like approx. line 519 in: 
https://github.com/QubesOS/qubes-core-admin/blob/master/qubes/storage/lvm.py
```
            cmd = ['create',
                   self.pool._pool_id,  # pylint: disable=protected-access
                   self._vid_import.split('/')[1],
                   str(src_volume.size)]
            yield from qubes_lvm_coro(cmd, self.log)
            src_path = src_volume.export()
            cmd = ['dd', 'if=' + src_path, 'of=/dev/' + self._vid_import,
                'conv=sparse', 'status=none']
            if not os.access('/dev/' + self._vid_import, os.W_OK) or \
                    not os.access(src_path, os.R_OK):
                cmd.insert(0, 'sudo')

            p = yield from asyncio.create_subprocess_exec(*cmd)
            yield from p.wait()
```

**Comment by brendanhoar** [comment]
replacing:
```
                'conv=sparse', 'status=none']
```
with:
```
                'conv=sparse', 'status=none', 'bs=128K']
```

...and rebooting, reduced the execution time of qvm-clone on this particular VM from 4m27s to 0m21s.

Brendan

**Comment by strugee** [comment]
Why bother with `dd` at all? `cat` or `cp` are probably [just as fast](https://unix.stackexchange.com/a/9492/29146) and simpler. There is also a `cp --sparse=always` flag. I guess it doesn't _really_ matter; maybe I'm bikeshedding a little...

**Comment by brendanhoar** [comment]
What is being cloned is a (virtual) device, not a file, so cp is not applicable.

**Comment by marmarek** [comment]
And preserving sparseness is important in the process, so `cat` is not good either.

**Comment by brendanhoar** [comment]
Just as a reminder, my suggested one line fix significantly improves the runtime of import_volume function (by a 10x speedup factor). This function is used when cloning VMs to a different LVM pool. I think it is also utilized when restoring VMs from a qubes backup to an LVM pool.

cp inherently uses this block size\*, but for lvm dd is being used. I recommend the bs=128k change be applied to lvm.py.

In addition, though I have not yet done performance testing, a similar change may help backups, which would be adding '-b 256' to the tar command parameters (256*512 bytes = 128k). The default is value is 20 (=10k), I believe.

B

\* An explantion of why cp was changed to utilize this blocksize as the standard blocksize is here (it essentially explains the `cp` source code comments): https://eklitzke.org/efficient-file-copying-on-linux

**Comment by brendanhoar** [comment]
Are these changes planned to be pushed to 4.0 testing as well?

**Comment by marmarek** [comment]
Yes, they are planned (and even to be included in 4.0.2-rc2).
